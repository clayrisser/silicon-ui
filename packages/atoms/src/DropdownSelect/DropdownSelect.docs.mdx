# Dropdown Select

> silicon ui dropdown select
