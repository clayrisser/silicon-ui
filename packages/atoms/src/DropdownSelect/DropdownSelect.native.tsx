import React, { FC } from 'react';
import styled, { StyledComponent } from '@emotion/primitives';
import { TouchableOpacity, Picker as NativePicker } from 'react-native';
import { createStyled } from '../styled';
import {
  background,
  border,
  color,
  compose,
  layout,
  position,
  shadow,
  space,
  typography
} from 'styled-system';
import useColor from '../hooks/useColor';
import {
  DropdownSelectProps,
  StyledDropdownSelectProps,
  splitProps
} from './dropdownSelectProps';

const StyledView: StyledComponent<
  StyledDropdownSelectProps,
  StyledDropdownSelectProps,
  any
> = styled.View(
  compose(
    background,
    border,
    color,
    layout,
    position,
    shadow,
    space,
    typography
  )
);

const StyledNativePicker = createStyled<DropdownSelectProps>(NativePicker, [
  background,
  border,
  layout,
  position,
  shadow,
  space
]);

const NativeBasePicker: FC<DropdownSelectProps> = (
  props: DropdownSelectProps
) => {
  const color = useColor(props);
  const {
    customDropdownSelectProps,
    nativeDropdownSelectProps,
    styledDropdownSelectProps,
    touchableOpacityProps
  } = splitProps({
    ...props,
    color
  });
  const children =
    typeof customDropdownSelectProps.children === 'string' ? (
      <NativePicker>{customDropdownSelectProps.children}</NativePicker>
    ) : (
      customDropdownSelectProps.children
    );

  return (
    <StyledNativePicker
      {...styledDropdownSelectProps}
      {...nativeDropdownSelectProps}
    >
      {children}
    </StyledNativePicker>
  );
};

NativeBasePicker.defaultProps = {
  // fontFamily: 'body',
  // fontWeight: 'body',
  // activeOpacity: 1,
  // autoContrast: false,
  // backgroundColor: 'background',
  children: <></>
  // fontSize: 0,
  // lineHeight: 'body'
};

export default NativeBasePicker;
