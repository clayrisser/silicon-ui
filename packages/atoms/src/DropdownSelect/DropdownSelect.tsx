import React, { FC } from 'react';
import styled, { StyledComponent } from '@emotion/styled';
import {
  background,
  border,
  color,
  compose,
  layout,
  position,
  shadow,
  space,
  typography
} from 'styled-system';
import useColor from '../hooks/useColor';
import {
  DropdownSelectProps,
  DetailedHTMLDropdownSelectProps,
  splitProps
} from './dropdownSelectProps';

const HTMLSelect: StyledComponent<
  DetailedHTMLDropdownSelectProps,
  DropdownSelectProps,
  object
> = styled.select(
  compose(
    background,
    border,
    color,
    layout,
    position,
    shadow,
    space,
    typography
  )
);

const DropdownSelect: FC<DropdownSelectProps> = (
  props: DropdownSelectProps
) => {
  const color = useColor(props);
  const {
    styledDropdownSelectProps,
    customDropdownSelectProps,
    touchableOpacityProps
  } = splitProps({
    ...props,
    color
  });

  return (
    <HTMLSelect
      {...customDropdownSelectProps}
      {...touchableOpacityProps}
      {...(styledDropdownSelectProps as any)}
      // onChange={(e: any) => console.log(e.target, e.target.value, 'detail')}
    />
  );
};

DropdownSelect.defaultProps = {
  activeOpacity: 1,
  backgroundColor: 'background',
  children: <></>,
  // value: '',
  fontFamily: 'body',
  fontSize: 0,
  fontWeight: 'body',
  lineHeight: 'body',
  width: '100%'
};

export default styled(DropdownSelect)`
  :active {
    opacity: ${({ activeOpacity }: DropdownSelectProps) => activeOpacity};
  }
`;
