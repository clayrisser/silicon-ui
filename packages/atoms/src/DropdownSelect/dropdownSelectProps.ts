import { ReactNode, SelectHTMLAttributes, DetailedHTMLProps } from 'react';
import { TouchableOpacityProps, ViewProps } from 'react-native';
import {
  BackgroundProps,
  BorderProps,
  ColorProps,
  LayoutProps,
  PositionProps,
  ShadowProps,
  SpaceProps,
  TypographyProps
} from 'styled-system';
import { Theme } from '../themes';

export type DetailedHTMLDropdownSelectProps = DetailedHTMLProps<
  SelectHTMLAttributes<HTMLSelectElement>,
  HTMLSelectElement
>;

export interface StyledDropdownSelectProps
  extends BackgroundProps,
    BorderProps,
    ColorProps,
    LayoutProps,
    PositionProps,
    ShadowProps,
    SpaceProps,
    TypographyProps {}

export interface NativeDropdownSelectProps extends ViewProps {
  selectedValue?: string;
}

export interface CustomDropdownSelectProps {
  autoContrast?: boolean | 'A' | 'AA' | 'AAA';
  children?: ReactNode;
  theme?: Theme;
  value?: string;
}

export interface DropdownSelectProps
  extends CustomDropdownSelectProps,
    NativeDropdownSelectProps,
    StyledDropdownSelectProps,
    TouchableOpacityProps {}

export const customDropdownSelectPropsKeys = new Set([
  'autoContrast',
  'children',
  'theme',
  'value'
]);

export const touchableOpacityPropsKeys = new Set([
  'activeOpacity',
  'delayLongPress',
  'delayPressIn',
  'delayPressOut',
  'disabled',
  'hitSlop',
  'onBlur',
  'onFocus',
  'onLayout',
  'onLongPress',
  'pressRetentionOffset',
  'testID'
]);

export const nativeDropdownSelectPropsKeys = new Set<string>([
  'onValueChange'
  // 'onPress',
  // 'onPressIn',
  // 'onPressOut'
]);

export interface SplitProps {
  customDropdownSelectProps: CustomDropdownSelectProps;
  nativeDropdownSelectProps: NativeDropdownSelectProps;
  styledDropdownSelectProps: StyledDropdownSelectProps;
  touchableOpacityProps: TouchableOpacityProps;
}

export function splitProps(props: DropdownSelectProps): SplitProps {
  const styledDropdownSelectProps: { [key: string]: any } = {};
  const customDropdownSelectProps: { [key: string]: any } = {};
  const touchableOpacityProps: { [key: string]: any } = {};
  const nativeDropdownSelectProps: { [key: string]: any } = {};
  Object.entries({ ...props }).forEach(([key, prop]: [string, any]) => {
    if (customDropdownSelectPropsKeys.has(key)) {
      customDropdownSelectProps[key] = prop;
    } else if (touchableOpacityPropsKeys.has(key)) {
      touchableOpacityProps[key] = prop;
    } else if (nativeDropdownSelectPropsKeys.has(key)) {
      nativeDropdownSelectProps[key] = prop;
    } else {
      styledDropdownSelectProps[key] = prop;
    }
  });
  return {
    customDropdownSelectProps,
    nativeDropdownSelectProps,
    styledDropdownSelectProps,
    touchableOpacityProps
  };
}
