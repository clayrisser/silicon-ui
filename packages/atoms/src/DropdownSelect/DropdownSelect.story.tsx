import React from 'react';
import { withKnobs, text, number, select } from '@storybook/addon-knobs';
import { Container, Header, Content, Form, Item, Picker } from 'native-base';

import DropdownSelect from './DropdownSelect';
import DropdownOption from '../DropdownOption/DropdownOption';
import Wrapper from '../../storybook/Wrapper';
import storiesOf from '../../storybook/storiesOf';
import themes from '../themes';
import withThemesProvider from '../../storybook/withThemesProvider';
import { Alert } from 'react-native';

storiesOf('Dropdown', module)
  .addDecorator(withKnobs)
  .addDecorator(withThemesProvider(themes))
  .addParameters({
    // docs: { page: docs },
    // jest: ['Button.test.tsx']
  })
  .add('with knobs', () => (
    <Wrapper>
      <DropdownSelect
        backgroundColor={text('backgroundColor', 'secondary')}
        // borderRadius={number('borderRadius', 0)}
        // height={200}
        // width={200}
        autoContrast={select<'A' | 'AA' | 'AAA'>(
          'autoContrast',
          {
            false: '' as 'A',
            A: 'A',
            AA: 'AA',
            AAA: 'AAA'
          },
          'AA'
        )}
        // selectedValue="Saab"
      >
        <DropdownOption value="Tesla" label="Tesla">
          Tesla
        </DropdownOption>
        <DropdownOption value="Volvo" label="Volvo">
          Volvo
        </DropdownOption>
        <DropdownOption value="Saab" label="Saab">
          Saab
        </DropdownOption>
        <DropdownOption value="Mercedes" label="Mercedes">
          Mercedes
        </DropdownOption>
        <DropdownOption value="Audi" label="Audi">
          Audi
        </DropdownOption>
        {/* <Picker.Item label="Wallet" value="key0" />
        <Picker.Item label="ATM Card" value="key1" />
        <Picker.Item label="Debit Card" value="key2" />
        <Picker.Item label="Credit Card" value="key3" />
        <Picker.Item label="Net Banking" value="key4" /> */}
      </DropdownSelect>
    </Wrapper>
  ));
