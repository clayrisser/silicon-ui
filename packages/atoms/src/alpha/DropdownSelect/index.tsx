import DropdownSelect from './DropdownSelect';
import { DropdownSelectProps } from './dropdownSelectProps';

export { DropdownSelectProps };
export default DropdownSelect;
