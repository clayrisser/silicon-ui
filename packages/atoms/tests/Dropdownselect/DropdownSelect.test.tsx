import React from 'react';
import renderer, { act, create } from 'react-test-renderer';
// import { render, fireEvent } from '@testing-library/react';
import DropdownSelect from '../../src/DropdownSelect';

describe('<DropdownSelect />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<DropdownSelect />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
describe('<DropdownOption withProps />', () => {
  it('renders  with all props correctly', () => {
    const value = 'Option';
    const display = 'Option';
    const disabled = true;
    const fontFamily = '';
    const backgroundColor = 'red';
    const fontWeight = 2;
    const fontSize = 3;
    const tree = renderer
      .create(
        <DropdownSelect
          value={value}
          display={display}
          backgroundColor={backgroundColor}
          disabled={disabled}
          fontFamily={fontFamily}
          fontWeight={fontWeight}
          fontSize={fontSize}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('<Dropdownselect/>', () => {
  it('renders with events', () => {
    const onChange = jest.fn();

    let component: any;
    act(() => {
      component = create(<DropdownSelect onChange={onChange} />);
    });

    const rootInstance: any = component.root;
    const dropdown = rootInstance.findAllByType(DropdownSelect);
    expect(dropdown.length).toBe(1);
    expect(dropdown[0].props.onChange()).toBe(undefined);
  });
});
