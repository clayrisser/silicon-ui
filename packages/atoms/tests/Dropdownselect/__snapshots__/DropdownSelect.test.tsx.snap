// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownSelect /> renders correctly 1`] = `
<select
  className="css-1uwwf6m"
  fontFamily="body"
  fontSize={0}
  fontWeight="body"
  onChange={[Function]}
  required={false}
  width="100%"
/>
`;
